### DISCUSSION TITLE
Auto-increment fields

### DISCUSSION THREAD

**Discussion by CodesInChaos** [geldata/gel#1709] [Q&A]
I understand the primary key is always a UUID in EdgeDB. But sometimes it'd be nice to have a smaller numeric secondary id. An auto-increment field combined with a unique (secondary) index on it would fit that goal.

Is there any way to define auto-increment fields in EdgeDB?

**Comment by elprans** [comment]
Yes:

```
scalar type TicketNo extending sequence;

type Ticket {
    property number -> TicketNo {
        constraint exclusive;
    }
}
```

**Reply by CodesInChaos** [reply]
The documentation doesn't mention this type in the [overview for scalar types](https://edgedb.com/docs/datamodel/scalars/index) where I looked for it. Though it apparently does include it on the detail page for numeric types. I think it'd be a good idea to list it on the overview page as well.

**Reply by 1st1** [reply]
@vpetrovykh Victor, let's add an example to the cheat sheet & make sure that `sequence` is documented.

**Comment by CodesInChaos** [comment]
When I define an auto-increment field in an abstract type, will the counter be shared between all concrete types derived from it or will they have their own counters? (assuming the unique constraint is delegated)

**Reply by elprans** [reply]
Shared. Sequences are bound to the scalar type, not the properties you're using them on.  There's no way to "delegate" the sequence currently.

**Reply by CodesInChaos** [reply]
1. Interesting. I didn't even consider that possibility. I only considered the per-concrete-property and per-abstract property options.
2. You should definitely mention that in the documentation.
3. Is the plain `sequence` type usable directly? If yes, does that mean all of its instances in the application share the same counter?

**Reply by elprans** [reply]
> Is the plain `sequence` type usable directly

No, it's declared as `abstract`

**Comment by CodesInChaos** [comment]
Conceptually coupling auto-increment primarily to a type seems wrong. It's a kind of default value, so I'd expect to specify where I specify a default value. While you could define a type for convenience, that can easily lead to unintended behaviour if such a type is ever used as a reference/foreign key (admittedly rare in EdgeDB due to its first class link support).

Also allowing auto-increment fields without a unique index seems looks like a foot gun to me. The values will usually be unique, but explicitly specified values or a reset sequence can mess up that guarantee.

**Reply by elprans** [reply]
`std::sequence` is conceptually similar (and is implemented in terms of) [PostgreSQL sequences](https://www.postgresql.org/docs/10/sql-createsequence.html), i.e. it's a standalone object that can be manipulated and queried.

> The values will usually be unique, but explicitly specified values or a reset sequence can mess up that guarantee.

`std::sequence` is a primitive, it'd be a significant orthogonality issue if it also imposed constraints on the users of it.  We can build on it by adding some special syntax for properties to achieve the effect you're suggesting.
